In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("Data Engineer Assignment.csv")

In [3]:
df.head()

provider_id  \
0  6873f0af240706d65514ec247c64f330   
1  a5a021d997269c4f1846903ff0f5d737   
2  77af6d6d6db530c389ec1014d557a01e   
3  7ea218c2b9a6dada27e0d5bdf5f896de   
4  0d7a9d9c00231885b413175980a9beca   

                                      detail             source  \
0                            {u'mme': False}      job_responded   
1  {u'lat': 18.9810108, u'long': 72.8187476}  location_tracking   
2                            {u'mme': False}    decline_request   
3                            {u'mme': False}    decline_request   
4  {u'lat': 12.9959477, u'long': 77.6527175}  location_tracking   

                   event_time  
0  2017-09-01 00:00:00.000000  
1  2017-09-01 00:00:02.000000  
2  2017-09-01 00:00:05.000000  
3  2017-09-01 00:00:12.000000  
4  2017-09-01 00:00:15.000000

In [4]:
df['event_time']=pd.to_datetime(df['event_time'])

In [5]:
df.head()

provider_id  \
0  6873f0af240706d65514ec247c64f330   
1  a5a021d997269c4f1846903ff0f5d737   
2  77af6d6d6db530c389ec1014d557a01e   
3  7ea218c2b9a6dada27e0d5bdf5f896de   
4  0d7a9d9c00231885b413175980a9beca   

                                      detail             source  \
0                            {u'mme': False}      job_responded   
1  {u'lat': 18.9810108, u'long': 72.8187476}  location_tracking   
2                            {u'mme': False}    decline_request   
3                            {u'mme': False}    decline_request   
4  {u'lat': 12.9959477, u'long': 77.6527175}  location_tracking   

           event_time  
0 2017-09-01 00:00:00  
1 2017-09-01 00:00:02  
2 2017-09-01 00:00:05  
3 2017-09-01 00:00:12  
4 2017-09-01 00:00:15

In [6]:
df.shape

(1945726, 4)

In [7]:
df[df.duplicated()].shape #check for duplicates 

(2696, 4)

In [8]:
df.drop_duplicates(inplace=True) #dropping all the duplicate rows
df[df.duplicated()].shape

(0, 4)

In [9]:
df.shape

(1943030, 4)

In [10]:
df.isnull().values.any() #check for any null values 

True

In [11]:
df['offline_status'] = df['detail'].str.contains("False")
df['offline_status'].replace(True,'offline', inplace=True)
df['offline_status'].replace(False, np.nan, inplace=True)

In [12]:
df['online_status'] = (df['detail'].str.contains("True")) | (df['source'].str.contains("Action on Job"))
df['online_status'].replace(True,'online', inplace=True)
df['online_status'].replace(False, np.nan, inplace=True)

In [13]:
df.head()

provider_id  \
0  6873f0af240706d65514ec247c64f330   
1  a5a021d997269c4f1846903ff0f5d737   
2  77af6d6d6db530c389ec1014d557a01e   
3  7ea218c2b9a6dada27e0d5bdf5f896de   
4  0d7a9d9c00231885b413175980a9beca   

                                      detail             source  \
0                            {u'mme': False}      job_responded   
1  {u'lat': 18.9810108, u'long': 72.8187476}  location_tracking   
2                            {u'mme': False}    decline_request   
3                            {u'mme': False}    decline_request   
4  {u'lat': 12.9959477, u'long': 77.6527175}  location_tracking   

           event_time offline_status online_status  
0 2017-09-01 00:00:00        offline           NaN  
1 2017-09-01 00:00:02            NaN           NaN  
2 2017-09-01 00:00:05        offline           NaN  
3 2017-09-01 00:00:12        offline           NaN  
4 2017-09-01 00:00:15            NaN           NaN

In [14]:
df.shape

(1943030, 6)

In [15]:
df.dropna(axis='index', how='all', subset=['offline_status', 'online_status'], inplace=True)
df.head()

provider_id           detail           source  \
0   6873f0af240706d65514ec247c64f330  {u'mme': False}    job_responded   
2   77af6d6d6db530c389ec1014d557a01e  {u'mme': False}  decline_request   
3   7ea218c2b9a6dada27e0d5bdf5f896de  {u'mme': False}  decline_request   
7   f1be9a4ebf57391f046464144a20fd36  {u'mme': False}    job_responded   
10  44f11b351f554bb6f564c9cd7c861645  {u'mme': False}    job_responded   

            event_time offline_status online_status  
0  2017-09-01 00:00:00        offline           NaN  
2  2017-09-01 00:00:05        offline           NaN  
3  2017-09-01 00:00:12        offline           NaN  
7  2017-09-01 00:00:27        offline           NaN  
10 2017-09-01 00:00:48        offline           NaN

In [16]:
df.shape

(855370, 6)

In [17]:
df['offline_status'].replace('offline', True, inplace=True)
df['offline_status'].replace(np.nan, False, inplace=True) 
df['online_status'].replace('online', True, inplace=True)
df['online_status'].replace(np.nan, False, inplace=True)
df.head()

provider_id           detail           source  \
0   6873f0af240706d65514ec247c64f330  {u'mme': False}    job_responded   
2   77af6d6d6db530c389ec1014d557a01e  {u'mme': False}  decline_request   
3   7ea218c2b9a6dada27e0d5bdf5f896de  {u'mme': False}  decline_request   
7   f1be9a4ebf57391f046464144a20fd36  {u'mme': False}    job_responded   
10  44f11b351f554bb6f564c9cd7c861645  {u'mme': False}    job_responded   

            event_time  offline_status  online_status  
0  2017-09-01 00:00:00            True          False  
2  2017-09-01 00:00:05            True          False  
3  2017-09-01 00:00:12            True          False  
7  2017-09-01 00:00:27            True          False  
10 2017-09-01 00:00:48            True          False

In [18]:
df.isnull().values.any() 

False

In [19]:
df['status'] = (df['online_status'] & ~df['offline_status']) | (df['online_status'] & df['offline_status']) #Boolean Expression derived from Boolean Algebra
df.head()

provider_id           detail           source  \
0   6873f0af240706d65514ec247c64f330  {u'mme': False}    job_responded   
2   77af6d6d6db530c389ec1014d557a01e  {u'mme': False}  decline_request   
3   7ea218c2b9a6dada27e0d5bdf5f896de  {u'mme': False}  decline_request   
7   f1be9a4ebf57391f046464144a20fd36  {u'mme': False}    job_responded   
10  44f11b351f554bb6f564c9cd7c861645  {u'mme': False}    job_responded   

            event_time  offline_status  online_status  status  
0  2017-09-01 00:00:00            True          False   False  
2  2017-09-01 00:00:05            True          False   False  
3  2017-09-01 00:00:12            True          False   False  
7  2017-09-01 00:00:27            True          False   False  
10 2017-09-01 00:00:48            True          False   False

In [20]:
df['status'].replace(True,'online', inplace=True)
df['status'].replace(False,'offline', inplace=True)
df.head()

provider_id           detail           source  \
0   6873f0af240706d65514ec247c64f330  {u'mme': False}    job_responded   
2   77af6d6d6db530c389ec1014d557a01e  {u'mme': False}  decline_request   
3   7ea218c2b9a6dada27e0d5bdf5f896de  {u'mme': False}  decline_request   
7   f1be9a4ebf57391f046464144a20fd36  {u'mme': False}    job_responded   
10  44f11b351f554bb6f564c9cd7c861645  {u'mme': False}    job_responded   

            event_time  offline_status  online_status   status  
0  2017-09-01 00:00:00            True          False  offline  
2  2017-09-01 00:00:05            True          False  offline  
3  2017-09-01 00:00:12            True          False  offline  
7  2017-09-01 00:00:27            True          False  offline  
10 2017-09-01 00:00:48            True          False  offline

In [21]:
df.drop(['detail','source','offline_status','online_status'], axis=1, inplace=True)
df.head()

provider_id          event_time   status
0   6873f0af240706d65514ec247c64f330 2017-09-01 00:00:00  offline
2   77af6d6d6db530c389ec1014d557a01e 2017-09-01 00:00:05  offline
3   7ea218c2b9a6dada27e0d5bdf5f896de 2017-09-01 00:00:12  offline
7   f1be9a4ebf57391f046464144a20fd36 2017-09-01 00:00:27  offline
10  44f11b351f554bb6f564c9cd7c861645 2017-09-01 00:00:48  offline

In [22]:
df['status'].value_counts()   #1.Num of online events and 2.Num of offline events

offline    614356
online     241014
Name: status, dtype: int64

### 1. Number of Events Offline = 6,14,356
### 2. Number of Events Online = 2,41,014

In [23]:
df['provider_id'].nunique()  #3a.Num of providers left after removing error rows

16283

### 3a. Number of Providers Left After Removing Error Rows = 16,283

In [24]:
df.head()

provider_id          event_time   status
0   6873f0af240706d65514ec247c64f330 2017-09-01 00:00:00  offline
2   77af6d6d6db530c389ec1014d557a01e 2017-09-01 00:00:05  offline
3   7ea218c2b9a6dada27e0d5bdf5f896de 2017-09-01 00:00:12  offline
7   f1be9a4ebf57391f046464144a20fd36 2017-09-01 00:00:27  offline
10  44f11b351f554bb6f564c9cd7c861645 2017-09-01 00:00:48  offline

In [25]:
df['hour'] = df['event_time'].dt.hour

In [26]:
df.tail()

provider_id          event_time   status  hour
1945714  9094a7169ad69befb34ed57640e12132 2017-09-30 23:58:47   online    23
1945715  cfdd7d7604c24c72d732a576e438a362 2017-09-30 23:58:48  offline    23
1945716  9094a7169ad69befb34ed57640e12132 2017-09-30 23:58:50  offline    23
1945718  cb817c584975076f22d4bd7d4b21b29e 2017-09-30 23:59:00  offline    23
1945720  ac2d43ab0b5574a2cc3034c4a71ca4dd 2017-09-30 23:59:06  offline    23

In [27]:
df.shape

(855370, 4)

In [28]:
df = df[(df['hour']>=0) & (df['hour']<19)]

In [29]:
df['provider_id'].nunique() #3b.Num of providers left after removing events on or after 7 PM

14914

### 3b. Number of Providers Left After Removing Events On or After 7 PM = 14,914

In [30]:
df.head()

provider_id          event_time   status  hour
0   6873f0af240706d65514ec247c64f330 2017-09-01 00:00:00  offline     0
2   77af6d6d6db530c389ec1014d557a01e 2017-09-01 00:00:05  offline     0
3   7ea218c2b9a6dada27e0d5bdf5f896de 2017-09-01 00:00:12  offline     0
7   f1be9a4ebf57391f046464144a20fd36 2017-09-01 00:00:27  offline     0
10  44f11b351f554bb6f564c9cd7c861645 2017-09-01 00:00:48  offline     0

In [31]:
df.drop('hour', axis=1, inplace=True)
df.head()

provider_id          event_time   status
0   6873f0af240706d65514ec247c64f330 2017-09-01 00:00:00  offline
2   77af6d6d6db530c389ec1014d557a01e 2017-09-01 00:00:05  offline
3   7ea218c2b9a6dada27e0d5bdf5f896de 2017-09-01 00:00:12  offline
7   f1be9a4ebf57391f046464144a20fd36 2017-09-01 00:00:27  offline
10  44f11b351f554bb6f564c9cd7c861645 2017-09-01 00:00:48  offline

In [32]:
provider_list = df['provider_id'].unique().tolist()

In [33]:
#print(provider_list)

In [34]:
time_init = pd.Timestamp('2017-09-01 19:00:00')
time_init

Timestamp('2017-09-01 19:00:00')

In [35]:
time_delta = pd.Timedelta('1 day')
time_delta

Timedelta('1 days 00:00:00')

In [36]:
time_list = []
time = time_init
for i in range(30):
    time_list.append(time)
    time += time_delta

In [37]:
#time_list

In [38]:
list_of_tuples = []

for provider in provider_list:
    for time in time_list:
        list_of_tuples.append((provider,time,'offline'))

In [39]:
df_EOD_offline = pd.DataFrame(list_of_tuples, columns = ['provider_id','event_time','status'])

#Reset to offline at EOD - For each provider for each date need added a row with timestamp "2017-09-xx 19:00:00" - 'provider_id' - 'offline'

In [40]:
df_total = pd.concat([df,df_EOD_offline], ignore_index=True)

In [41]:
df_total.tail()

provider_id          event_time   status
1142211  67c693c90bc9793463ef0bde01a627e5 2017-09-26 19:00:00  offline
1142212  67c693c90bc9793463ef0bde01a627e5 2017-09-27 19:00:00  offline
1142213  67c693c90bc9793463ef0bde01a627e5 2017-09-28 19:00:00  offline
1142214  67c693c90bc9793463ef0bde01a627e5 2017-09-29 19:00:00  offline
1142215  67c693c90bc9793463ef0bde01a627e5 2017-09-30 19:00:00  offline

In [42]:
df_total = df_total.set_index('event_time')
df_total.head()

provider_id   status
event_time                                                    
2017-09-01 00:00:00  6873f0af240706d65514ec247c64f330  offline
2017-09-01 00:00:05  77af6d6d6db530c389ec1014d557a01e  offline
2017-09-01 00:00:12  7ea218c2b9a6dada27e0d5bdf5f896de  offline
2017-09-01 00:00:27  f1be9a4ebf57391f046464144a20fd36  offline
2017-09-01 00:00:48  44f11b351f554bb6f564c9cd7c861645  offline

In [43]:
df_total['time'] = df_total.index
df_total.head()

provider_id   status  \
event_time                                                       
2017-09-01 00:00:00  6873f0af240706d65514ec247c64f330  offline   
2017-09-01 00:00:05  77af6d6d6db530c389ec1014d557a01e  offline   
2017-09-01 00:00:12  7ea218c2b9a6dada27e0d5bdf5f896de  offline   
2017-09-01 00:00:27  f1be9a4ebf57391f046464144a20fd36  offline   
2017-09-01 00:00:48  44f11b351f554bb6f564c9cd7c861645  offline   

                                   time  
event_time                               
2017-09-01 00:00:00 2017-09-01 00:00:00  
2017-09-01 00:00:05 2017-09-01 00:00:05  
2017-09-01 00:00:12 2017-09-01 00:00:12  
2017-09-01 00:00:27 2017-09-01 00:00:27  
2017-09-01 00:00:48 2017-09-01 00:00:48

In [44]:
df_total.sort_values(['provider_id','time'], inplace=True)

In [45]:
df_total.head()

provider_id   status  \
event_time                                                       
2017-09-01 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-02 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-03 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-04 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-05 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   

                                   time  
event_time                               
2017-09-01 19:00:00 2017-09-01 19:00:00  
2017-09-02 19:00:00 2017-09-02 19:00:00  
2017-09-03 19:00:00 2017-09-03 19:00:00  
2017-09-04 19:00:00 2017-09-04 19:00:00  
2017-09-05 19:00:00 2017-09-05 19:00:00

In [46]:
df_total['date'] = df_total['time'].map(pd.Timestamp.date)
df_total['hour startime'] = df_total['time'].dt.hour
df_total['hour endtime'] = df_total['hour startime'] + 1

In [47]:
df_total.head()

provider_id   status  \
event_time                                                       
2017-09-01 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-02 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-03 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-04 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-05 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-01 19:00:00 2017-09-01 19:00:00  2017-09-01             19   
2017-09-02 19:00:00 2017-09-02 19:00:00  2017-09-02             19   
2017-09-03 19:00:00 2017-09-03 19:00:00  2017-09-03             19   
2017-09-04 19:00:00 2017-09-04 19:00:00  2017-09-04             19   
2017-09-05 19:00:00 2017-09-05 19:00:00  2017-09-05             19   

                     hour endtime  
event_time                         
2017-09-01 19:00:00            20  
2017-09-02 19:00:00            20  
2017-09-03 19:00:00            20  
2017-09-04 19:00:00            20  
2017-09-05 19:00:00            20

In [48]:
df_total['delta'] = df_total['time']-df_total['time'].shift()

In [49]:
df_total.head()

provider_id   status  \
event_time                                                       
2017-09-01 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-02 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-03 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-04 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-05 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-01 19:00:00 2017-09-01 19:00:00  2017-09-01             19   
2017-09-02 19:00:00 2017-09-02 19:00:00  2017-09-02             19   
2017-09-03 19:00:00 2017-09-03 19:00:00  2017-09-03             19   
2017-09-04 19:00:00 2017-09-04 19:00:00  2017-09-04             19   
2017-09-05 19:00:00 2017-09-05 19:00:00  2017-09-05             19   

                     hour endtime  delta  
event_time                                
2017-09-01 19:00:00            20    NaT  
2017-09-02 19:00:00            20 1 days  
2017-09-03 19:00:00            20 1 days  
2017-09-04 19:00:00            20 1 days  
2017-09-05 19:00:00            20 1 days

In [50]:
df_total['delta']=pd.to_timedelta(df_total['delta'], errors='coerce')

In [51]:
df_total.isnull().values.any()

True

In [52]:
df_total.shape

(1142216, 7)

In [53]:
df_total.dropna(inplace=True)

In [54]:
df_total.shape

(1142215, 7)

In [55]:
df_total['delta1'] = df_total['delta'].mask(df_total['status'].shift(-1)=='offline',0)  
#time where the provider is online

In [56]:
df_total.head()

provider_id   status  \
event_time                                                       
2017-09-02 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-03 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-04 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-05 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-06 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-02 19:00:00 2017-09-02 19:00:00  2017-09-02             19   
2017-09-03 19:00:00 2017-09-03 19:00:00  2017-09-03             19   
2017-09-04 19:00:00 2017-09-04 19:00:00  2017-09-04             19   
2017-09-05 19:00:00 2017-09-05 19:00:00  2017-09-05             19   
2017-09-06 19:00:00 2017-09-06 19:00:00  2017-09-06             19   

                     hour endtime  delta delta1  
event_time                                       
2017-09-02 19:00:00            20 1 days      0  
2017-09-03 19:00:00            20 1 days      0  
2017-09-04 19:00:00            20 1 days      0  
2017-09-05 19:00:00            20 1 days      0  
2017-09-06 19:00:00            20 1 days      0

In [57]:
#df_total.dtypes

In [58]:
df_total['delta1'] = pd.to_timedelta(df_total['delta1']) #conversion from object datatype to timedelta64[ns]

In [59]:
df_total['delta1'] = df_total['delta1'].dt.total_seconds() #conversion of delta1 to seconds online

In [60]:
df_total.head(10)

provider_id   status  \
event_time                                                       
2017-09-02 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-03 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-04 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-05 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-06 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-07 00:30:46  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-07 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-08 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-09 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-10 19:00:00  00020da44b0a44e867de5f26c69886ac  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-02 19:00:00 2017-09-02 19:00:00  2017-09-02             19   
2017-09-03 19:00:00 2017-09-03 19:00:00  2017-09-03             19   
2017-09-04 19:00:00 2017-09-04 19:00:00  2017-09-04             19   
2017-09-05 19:00:00 2017-09-05 19:00:00  2017-09-05             19   
2017-09-06 19:00:00 2017-09-06 19:00:00  2017-09-06             19   
2017-09-07 00:30:46 2017-09-07 00:30:46  2017-09-07              0   
2017-09-07 19:00:00 2017-09-07 19:00:00  2017-09-07             19   
2017-09-08 19:00:00 2017-09-08 19:00:00  2017-09-08             19   
2017-09-09 19:00:00 2017-09-09 19:00:00  2017-09-09             19   
2017-09-10 19:00:00 2017-09-10 19:00:00  2017-09-10             19   

                     hour endtime           delta  delta1  
event_time                                                 
2017-09-02 19:00:00            20 1 days 00:00:00     0.0  
2017-09-03 19:00:00            20 1 days 00:00:00     0.0  
2017-09-04 19:00:00            20 1 days 00:00:00     0.0  
2017-09-05 19:00:00            20 1 days 00:00:00     0.0  
2017-09-06 19:00:00            20 1 days 00:00:00     0.0  
2017-09-07 00:30:46             1 0 days 05:30:46     0.0  
2017-09-07 19:00:00            20 0 days 18:29:14     0.0  
2017-09-08 19:00:00            20 1 days 00:00:00     0.0  
2017-09-09 19:00:00            20 1 days 00:00:00     0.0  
2017-09-10 19:00:00            20 1 days 00:00:00     0.0

In [61]:
df_total[df_total['delta1']>0].head()

provider_id   status  \
event_time                                                       
2017-09-03 19:00:00  0037dfffff8b03bbdf366a263735e84b  offline   
2017-09-04 08:00:19  0037dfffff8b03bbdf366a263735e84b   online   
2017-09-04 08:44:36  0037dfffff8b03bbdf366a263735e84b  offline   
2017-09-04 09:06:12  0037dfffff8b03bbdf366a263735e84b   online   
2017-09-04 12:45:46  0037dfffff8b03bbdf366a263735e84b  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-03 19:00:00 2017-09-03 19:00:00  2017-09-03             19   
2017-09-04 08:00:19 2017-09-04 08:00:19  2017-09-04              8   
2017-09-04 08:44:36 2017-09-04 08:44:36  2017-09-04              8   
2017-09-04 09:06:12 2017-09-04 09:06:12  2017-09-04              9   
2017-09-04 12:45:46 2017-09-04 12:45:46  2017-09-04             12   

                     hour endtime           delta   delta1  
event_time                                                  
2017-09-03 19:00:00            20 1 days 00:00:00  86400.0  
2017-09-04 08:00:19             9 0 days 13:00:19  46819.0  
2017-09-04 08:44:36             9 0 days 00:14:57    897.0  
2017-09-04 09:06:12            10 0 days 00:21:36   1296.0  
2017-09-04 12:45:46            13 0 days 00:21:59   1319.0

In [62]:
df_new = df_total[df_total['hour startime']!=19]

In [63]:
df_new.head(20)

provider_id   status  \
event_time                                                       
2017-09-07 00:30:46  00020da44b0a44e867de5f26c69886ac  offline   
2017-09-01 17:37:59  0002beae790876e71dfb99a904563c30  offline   
2017-09-01 17:38:05  0002beae790876e71dfb99a904563c30  offline   
2017-09-01 17:38:12  0002beae790876e71dfb99a904563c30  offline   
2017-09-01 17:38:18  0002beae790876e71dfb99a904563c30  offline   
2017-09-02 16:52:19  0002beae790876e71dfb99a904563c30  offline   
2017-09-02 16:52:33  0002beae790876e71dfb99a904563c30  offline   
2017-09-03 11:39:17  0002beae790876e71dfb99a904563c30  offline   
2017-09-03 11:39:22  0002beae790876e71dfb99a904563c30  offline   
2017-09-03 14:16:03  0002beae790876e71dfb99a904563c30  offline   
2017-09-03 14:16:07  0002beae790876e71dfb99a904563c30  offline   
2017-09-03 14:16:16  0002beae790876e71dfb99a904563c30  offline   
2017-09-05 14:33:49  0002beae790876e71dfb99a904563c30  offline   
2017-09-05 14:34:29  0002beae790876e71dfb99a904563c30  offline   
2017-09-05 14:34:36  0002beae790876e71dfb99a904563c30  offline   
2017-09-01 08:05:45  0005d4e14f83f51f1bbbd28ea6d8b7e7  offline   
2017-09-01 15:27:11  0005d4e14f83f51f1bbbd28ea6d8b7e7  offline   
2017-09-01 15:48:40  0005d4e14f83f51f1bbbd28ea6d8b7e7  offline   
2017-09-01 16:09:18  0005d4e14f83f51f1bbbd28ea6d8b7e7  offline   
2017-09-01 16:27:58  0005d4e14f83f51f1bbbd28ea6d8b7e7  offline   

                                   time        date  hour startime  \
event_time                                                           
2017-09-07 00:30:46 2017-09-07 00:30:46  2017-09-07              0   
2017-09-01 17:37:59 2017-09-01 17:37:59  2017-09-01             17   
2017-09-01 17:38:05 2017-09-01 17:38:05  2017-09-01             17   
2017-09-01 17:38:12 2017-09-01 17:38:12  2017-09-01             17   
2017-09-01 17:38:18 2017-09-01 17:38:18  2017-09-01             17   
2017-09-02 16:52:19 2017-09-02 16:52:19  2017-09-02             16   
2017-09-02 16:52:33 2017-09-02 16:52:33  2017-09-02             16   
2017-09-03 11:39:17 2017-09-03 11:39:17  2017-09-03             11   
2017-09-03 11:39:22 2017-09-03 11:39:22  2017-09-03             11   
2017-09-03 14:16:03 2017-09-03 14:16:03  2017-09-03             14   
2017-09-03 14:16:07 2017-09-03 14:16:07  2017-09-03             14   
2017-09-03 14:16:16 2017-09-03 14:16:16  2017-09-03             14   
2017-09-05 14:33:49 2017-09-05 14:33:49  2017-09-05             14   
2017-09-05 14:34:29 2017-09-05 14:34:29  2017-09-05             14   
2017-09-05 14:34:36 2017-09-05 14:34:36  2017-09-05             14   
2017-09-01 08:05:45 2017-09-01 08:05:45  2017-09-01              8   
2017-09-01 15:27:11 2017-09-01 15:27:11  2017-09-01             15   
2017-09-01 15:48:40 2017-09-01 15:48:40  2017-09-01             15   
2017-09-01 16:09:18 2017-09-01 16:09:18  2017-09-01             16   
2017-09-01 16:27:58 2017-09-01 16:27:58  2017-09-01             16   

                     hour endtime              delta  delta1  
event_time                                                    
2017-09-07 00:30:46             1    0 days 05:30:46     0.0  
2017-09-01 17:37:59            18 -30 days +22:37:59     0.0  
2017-09-01 17:38:05            18    0 days 00:00:06     0.0  
2017-09-01 17:38:12            18    0 days 00:00:07     0.0  
2017-09-01 17:38:18            18    0 days 00:00:06     0.0  
2017-09-02 16:52:19            17    0 days 21:52:19     0.0  
2017-09-02 16:52:33            17    0 days 00:00:14     0.0  
2017-09-03 11:39:17            12    0 days 16:39:17     0.0  
2017-09-03 11:39:22            12    0 days 00:00:05     0.0  
2017-09-03 14:16:03            15    0 days 02:36:41     0.0  
2017-09-03 14:16:07            15    0 days 00:00:04     0.0  
2017-09-03 14:16:16            15    0 days 00:00:09     0.0  
2017-09-05 14:33:49            15    0 days 19:33:49     0.0  
2017-09-05 14:34:29            15    0 days 00:00:40     0.0  
2017-09-05 14:34:36            15    0 d

In [64]:
g = df_new.groupby(['provider_id','date','hour startime','hour endtime'])

In [65]:
final_df = g.sum()

In [66]:
final_df.head() 

delta1
provider_id                      date       hour startime hour endtime        
00020da44b0a44e867de5f26c69886ac 2017-09-07 0             1                0.0
0002beae790876e71dfb99a904563c30 2017-09-01 17            18               0.0
                                 2017-09-02 16            17               0.0
                                 2017-09-03 11            12               0.0
                                            14            15               0.0

In [67]:
final_df.shape #4.Final output dataframe contains 316223 rows

(316223, 1)

### 4. Final output dataframe contains 3,16,223 rows

In [68]:
final_df[final_df['delta1']>0] #5.Num of rows with seconds online>0 = 95,888

delta1
provider_id                      date       hour startime hour endtime         
0037dfffff8b03bbdf366a263735e84b 2017-09-04 8             9             47716.0
                                            9             10             1296.0
                                            12            13             1329.0
                                            17            18            17695.0
                                 2017-09-05 9             10            50435.0
                                            11            12             6287.0
                                            13            14             7285.0
                                            15            16               50.0
                                            17            18             6052.0
                                 2017-09-06 10            11             9318.0
                                            11            12             4913.0
                                            12            13               27.0
                                            14            15             7073.0
                                            15            16              915.0
                                            16            17               89.0
                                            17            18             5091.0
                                 2017-09-07 7             8             46111.0
                                            9             10               38.0
                                            12            13            10128.0
                                            13            14             4788.0
                                            14            15               80.0
                                            16            17             6035.0
                                            17            18             5044.0
                                 2017-09-08 10            11            55795.0
                                            12            13             6144.0
                                            13            14             2885.0
                                            14            15             5729.0
                                 2017-09-09 11            12            57710.0
                                            15            16            15029.0
                                 2017-09-10 9             10            52396.0
...                                                                         ...
ffc01e5fd6946d3110c2e8b99aa5c469 2017-09-25 7             8              2256.0
                                            8             9              3719.0
                                            11            12             8853.0
                                            12            13             3352.0
                                 2017-09-26 15            16            74860.0
                                 2017-09-27 8             9             49244.0
                                            13            14            16250.0
                                 2017-09-28 8             9             48136.0
                                            9             10             4469.0
                                            11            12             4931.0
                                            14            15            11896.0
                                            16            17             5534.0
                                            17            18             2621.0
                                 2017-09-29 7             8             43402.0
                                            9             10             8386.0
                                            16            17            23285.0
                                            17            18             3046.0
                                 2017-09-30 7             8             4

### 5. Number of rows with seconds online greater than 0 = 95,888